File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/LibrarySourceStateStore.kt
package com.pillyliu.pinprofandroid.library

import android.content.Context
import androidx.core.content.edit
import org.json.JSONArray
import org.json.JSONObject

internal data class LibrarySourceState(
    val enabledSourceIds: List<String> = emptyList(),
    val pinnedSourceIds: List<String> = emptyList(),
    val selectedSourceId: String? = null,
    val selectedSortBySource: Map<String, String> = emptyMap(),
    val selectedBankBySource: Map<String, Int> = emptyMap(),
)

internal object LibrarySourceStateStore {
    private const val PREFS_NAME = "practice-upgrade-state-v2"
    private const val STATE_KEY = "pinball-library-source-state-v1"
    const val MAX_PINNED_SOURCES = 10

    fun hasPersistedState(context: Context): Boolean {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        return prefs.contains(STATE_KEY)
    }

    fun load(context: Context): LibrarySourceState {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val raw = prefs.getString(STATE_KEY, null) ?: return LibrarySourceState()
        return runCatching {
            val root = JSONObject(raw)
            val decoded = LibrarySourceState(
                enabledSourceIds = root.optJSONArray("enabledSourceIds").toStringList(),
                pinnedSourceIds = root.optJSONArray("pinnedSourceIds").toStringList(),
                selectedSourceId = root.optString("selectedSourceId").trim().ifBlank { null },
                selectedSortBySource = root.optJSONObject("selectedSortBySource").toStringMap(),
                selectedBankBySource = root.optJSONObject("selectedBankBySource").toIntMap(),
            )
            val normalized = normalize(decoded)
            if (normalized != decoded) {
                save(context, normalized)
            }
            normalized
        }.getOrDefault(LibrarySourceState())
    }

    fun save(context: Context, state: LibrarySourceState) {
        val normalized = normalize(state)
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val json = JSONObject().apply {
            put("enabledSourceIds", JSONArray().apply { normalized.enabledSourceIds.forEach(::put) })
            put("pinnedSourceIds", JSONArray().apply { normalized.pinnedSourceIds.forEach(::put) })
            put("selectedSourceId", normalized.selectedSourceId)
            put("selectedSortBySource", JSONObject().apply {
                normalized.selectedSortBySource.forEach { (key, value) -> put(key, value) }
            })
            put("selectedBankBySource", JSONObject().apply {
                normalized.selectedBankBySource.forEach { (key, value) -> put(key, value) }
            })
        }
        prefs.edit { putString(STATE_KEY, json.toString()) }
    }

    fun synchronize(context: Context, payloadSources: List<LibrarySource>): LibrarySourceState {
        val validIds = payloadSources.map { it.id }.toSet()
        var state = load(context)
        state = state.copy(
            enabledSourceIds = filteredKnownIds(state.enabledSourceIds, validIds),
            pinnedSourceIds = filteredKnownIds(state.pinnedSourceIds, validIds).take(MAX_PINNED_SOURCES),
            selectedSourceId = canonicalLibrarySourceId(state.selectedSourceId)?.takeIf { validIds.contains(it) },
            selectedSortBySource = state.selectedSortBySource.mapNotNullKeys(::canonicalLibrarySourceId).filterKeys { validIds.contains(it) },
            selectedBankBySource = state.selectedBankBySource.mapNotNullKeys(::canonicalLibrarySourceId).filterKeys { validIds.contains(it) },
        )
        if (!hasPersistedState(context)) {
            val seededIds = DEFAULT_SEEDED_LIBRARY_SOURCE_IDS.filter { validIds.contains(it) }
            if (seededIds.isNotEmpty()) {
                state = state.copy(
                    enabledSourceIds = seededIds,
                    pinnedSourceIds = seededIds.take(MAX_PINNED_SOURCES),
                    selectedSourceId = seededIds.firstOrNull(),
                )
            }
        }
        save(context, state)
        return state
    }

    fun upsertSource(context: Context, id: String, enable: Boolean = true, pinIfPossible: Boolean = true) {
        val canonicalId = canonicalLibrarySourceId(id) ?: return
        val current = load(context)
        val enabled = current.enabledSourceIds.toMutableList()
        val pinned = current.pinnedSourceIds.toMutableList()
        if (enable && !enabled.contains(canonicalId)) {
            enabled += canonicalId
        }
        if (pinIfPossible && !pinned.contains(canonicalId) && pinned.size < MAX_PINNED_SOURCES) {
            pinned += canonicalId
        }
        save(context, current.copy(enabledSourceIds = enabled, pinnedSourceIds = pinned))
    }

    fun setEnabled(context: Context, sourceId: String, isEnabled: Boolean) {
        val canonicalId = canonicalLibrarySourceId(sourceId) ?: return
        val current = load(context)
        val enabled = current.enabledSourceIds.toMutableList()
        val pinned = current.pinnedSourceIds.toMutableList()
        val selected = current.selectedSourceId
        if (isEnabled) {
            if (!enabled.contains(canonicalId)) enabled += canonicalId
            save(context, current.copy(enabledSourceIds = enabled))
            return
        }
        enabled.removeAll { it == canonicalId }
        pinned.removeAll { it == canonicalId }
        save(
            context,
            current.copy(
                enabledSourceIds = enabled,
                pinnedSourceIds = pinned,
                selectedSourceId = if (selected == canonicalId) null else selected,
            ),
        )
    }

    fun setPinned(context: Context, sourceId: String, isPinned: Boolean): Boolean {
        val canonicalId = canonicalLibrarySourceId(sourceId) ?: return false
        val current = load(context)
        val enabled = current.enabledSourceIds.toMutableList()
        val pinned = current.pinnedSourceIds.toMutableList()
        if (isPinned) {
            if (pinned.contains(canonicalId)) return true
            if (pinned.size >= MAX_PINNED_SOURCES) return false
            if (!enabled.contains(canonicalId)) enabled += canonicalId
            pinned += canonicalId
        } else {
            pinned.removeAll { it == canonicalId }
        }
        save(context, current.copy(enabledSourceIds = enabled, pinnedSourceIds = pinned))
        return true
    }

    fun setSelectedSource(context: Context, sourceId: String?) {
        save(context, load(context).copy(selectedSourceId = canonicalLibrarySourceId(sourceId)))
    }

    fun setSelectedSort(context: Context, sourceId: String, sortName: String) {
        val canonicalId = canonicalLibrarySourceId(sourceId) ?: return
        val current = load(context)
        val next = current.selectedSortBySource.toMutableMap()
        next[canonicalId] = sortName
        save(context, current.copy(selectedSortBySource = next))
    }

    fun setSelectedBank(context: Context, sourceId: String, bank: Int?) {
        val canonicalId = canonicalLibrarySourceId(sourceId) ?: return
        val current = load(context)
        val next = current.selectedBankBySource.toMutableMap()
        if (bank == null) next.remove(canonicalId) else next[canonicalId] = bank
        save(context, current.copy(selectedBankBySource = next))
    }

    fun removeSourcePreferences(context: Context, sourceId: String) {
        val canonicalId = canonicalLibrarySourceId(sourceId) ?: return
        val current = load(context)
        save(
            context,
            current.copy(
                enabledSourceIds = current.enabledSourceIds.filterNot { it == canonicalId },
                pinnedSourceIds = current.pinnedSourceIds.filterNot { it == canonicalId },
                selectedSourceId = current.selectedSourceId?.takeUnless { it == canonicalId },
                selectedSortBySource = current.selectedSortBySource.filterKeys { it != canonicalId },
                selectedBankBySource = current.selectedBankBySource.filterKeys { it != canonicalId },
            ),
        )
    }

    private fun normalize(state: LibrarySourceState): LibrarySourceState =
        LibrarySourceState(
            enabledSourceIds = state.enabledSourceIds.mapNotNull(::canonicalLibrarySourceId).distinct(),
            pinnedSourceIds = state.pinnedSourceIds.mapNotNull(::canonicalLibrarySourceId).distinct(),
            selectedSourceId = canonicalLibrarySourceId(state.selectedSourceId),
            selectedSortBySource = state.selectedSortBySource.mapNotNullKeys(::canonicalLibrarySourceId),
            selectedBankBySource = state.selectedBankBySource.mapNotNullKeys(::canonicalLibrarySourceId),
        )

    private fun filteredKnownIds(ids: List<String>, validIds: Set<String>): List<String> {
        val seen = LinkedHashSet<String>()
        return ids.mapNotNull(::canonicalLibrarySourceId).filter { id -> validIds.contains(id) && seen.add(id) }
    }
}

File: Pinball App 2/Pinball App 2/library/LibrarySourceStateStore.swift
import Foundation

struct PinballLibrarySourceState: Codable, Equatable {
    var enabledSourceIDs: [String]
    var pinnedSourceIDs: [String]
    var selectedSourceID: String?
    var selectedSortBySource: [String: String]
    var selectedBankBySource: [String: Int]

    static let empty = PinballLibrarySourceState(
        enabledSourceIDs: [],
        pinnedSourceIDs: [],
        selectedSourceID: nil,
        selectedSortBySource: [:],
        selectedBankBySource: [:]
    )
}

enum PinballLibrarySourceStateStore {
    private static let defaultsKey = "pinball-library-source-state-v1"
    static let maxPinnedSources = 10

    static func hasPersistedState() -> Bool {
        UserDefaults.standard.data(forKey: defaultsKey) != nil
    }

    static func load() -> PinballLibrarySourceState {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey) else {
            return .empty
        }
        if let state = try? JSONDecoder().decode(PinballLibrarySourceState.self, from: data) {
            let migrated = normalized(state)
            if migrated != state {
                save(migrated)
            }
            return migrated
        }
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return .empty
        }
        let migrated = normalized(
            PinballLibrarySourceState(
                enabledSourceIDs: (root["enabledSourceIDs"] as? [Any])?.compactMap { canonicalLibrarySourceID(String(describing: $0)) } ?? [],
                pinnedSourceIDs: (root["pinnedSourceIDs"] as? [Any])?.compactMap { canonicalLibrarySourceID(String(describing: $0)) } ?? [],
                selectedSourceID: canonicalLibrarySourceID(root["selectedSourceID"] as? String),
                selectedSortBySource: normalizeStringMap(root["selectedSortBySource"] as? [String: Any]),
                selectedBankBySource: normalizeIntMap(root["selectedBankBySource"] as? [String: Any])
            )
        )
        save(migrated)
        return migrated
    }

    static func save(_ state: PinballLibrarySourceState) {
        guard let data = try? JSONEncoder().encode(state) else { return }
        UserDefaults.standard.set(data, forKey: defaultsKey)
    }

    static func synchronize(with payloadSources: [PinballLibrarySource]) -> PinballLibrarySourceState {
        let validIDs = Set(payloadSources.map(\.id))
        var state = load()
        state.enabledSourceIDs = filteredKnownIDs(state.enabledSourceIDs, validIDs: validIDs)
        state.pinnedSourceIDs = Array(filteredKnownIDs(state.pinnedSourceIDs, validIDs: validIDs).prefix(maxPinnedSources))
        if !hasPersistedState() {
            let seededIDs = defaultSeededLibrarySourceIDs.filter { validIDs.contains($0) }
            if !seededIDs.isEmpty {
                state.enabledSourceIDs = seededIDs
                state.pinnedSourceIDs = Array(seededIDs.prefix(maxPinnedSources))
                state.selectedSourceID = seededIDs.first
            }
        }

        if let selectedSourceID = canonicalLibrarySourceID(state.selectedSourceID), validIDs.contains(selectedSourceID) {
            state.selectedSourceID = selectedSourceID
        } else {
            state.selectedSourceID = nil
        }

        state.selectedSortBySource = Dictionary(
            uniqueKeysWithValues: dedupedPairs(state.selectedSortBySource.compactMap { key, value in
                guard let canonicalKey = canonicalLibrarySourceID(key), validIDs.contains(canonicalKey) else { return nil }
                return (canonicalKey, value)
            })
        )
        state.selectedBankBySource = Dictionary(
            uniqueKeysWithValues: dedupedPairs(state.selectedBankBySource.compactMap { key, value in
                guard let canonicalKey = canonicalLibrarySourceID(key), validIDs.contains(canonicalKey) else { return nil }
                return (canonicalKey, value)
            })
        )
        save(state)
        return state
    }

    static func upsertSource(id: String, enable: Bool = true, pinIfPossible: Bool = true) {
        guard let id = canonicalLibrarySourceID(id) else { return }
        var state = load()
        if enable, !state.enabledSourceIDs.contains(id) {
            state.enabledSourceIDs.append(id)
        }
        if pinIfPossible, !state.pinnedSourceIDs.contains(id), state.pinnedSourceIDs.count < maxPinnedSources {
            state.pinnedSourceIDs.append(id)
        }
        save(state)
    }

    static func setEnabled(sourceID: String, isEnabled: Bool) {
        guard let sourceID = canonicalLibrarySourceID(sourceID) else { return }
        var state = load()
        if isEnabled {
            if !state.enabledSourceIDs.contains(sourceID) {
                state.enabledSourceIDs.append(sourceID)
            }
        } else {
            state.enabledSourceIDs.removeAll { $0 == sourceID }
            state.pinnedSourceIDs.removeAll { $0 == sourceID }
            if state.selectedSourceID == sourceID {
                state.selectedSourceID = nil
            }
        }
        save(state)
    }

    static func setPinned(sourceID: String, isPinned: Bool) -> Bool {
        guard let sourceID = canonicalLibrarySourceID(sourceID) else { return false }
        var state = load()
        if isPinned {
            if state.pinnedSourceIDs.contains(sourceID) {
                return true
            }
            guard state.pinnedSourceIDs.count < maxPinnedSources else {
                return false
            }
            if !state.enabledSourceIDs.contains(sourceID) {
                state.enabledSourceIDs.append(sourceID)
            }
            state.pinnedSourceIDs.append(sourceID)
        } else {
            state.pinnedSourceIDs.removeAll { $0 == sourceID }
        }
        save(state)
        return true
    }

    static func setSelectedSourceID(_ sourceID: String?) {
        var state = load()
        state.selectedSourceID = canonicalLibrarySourceID(sourceID)
        save(state)
    }

    static func setSelectedSort(sourceID: String, sortName: String) {
        guard let sourceID = canonicalLibrarySourceID(sourceID) else { return }
        var state = load()
        state.selectedSortBySource[sourceID] = sortName
        save(state)
    }

    static func setSelectedBank(sourceID: String, bank: Int?) {
        guard let sourceID = canonicalLibrarySourceID(sourceID) else { return }
        var state = load()
        if let bank {
            state.selectedBankBySource[sourceID] = bank
        } else {
            state.selectedBankBySource.removeValue(forKey: sourceID)
        }
        save(state)
    }

    static func removeSourcePreferences(sourceID: String) {
        guard let sourceID = canonicalLibrarySourceID(sourceID) else { return }
        var state = load()
        state.enabledSourceIDs.removeAll { $0 == sourceID }
        state.pinnedSourceIDs.removeAll { $0 == sourceID }
        if state.selectedSourceID == sourceID {
            state.selectedSourceID = nil
        }
        state.selectedSortBySource.removeValue(forKey: sourceID)
        state.selectedBankBySource.removeValue(forKey: sourceID)
        save(state)
    }

    private static func filteredKnownIDs(_ ids: [String], validIDs: Set<String>) -> [String] {
        var seen = Set<String>()
        return ids.compactMap(canonicalLibrarySourceID).filter { id in
            validIDs.contains(id) && seen.insert(id).inserted
        }
    }

    private static func normalized(_ state: PinballLibrarySourceState) -> PinballLibrarySourceState {
        PinballLibrarySourceState(
            enabledSourceIDs: Array(NSOrderedSet(array: state.enabledSourceIDs.compactMap(canonicalLibrarySourceID))) as? [String] ?? [],
            pinnedSourceIDs: Array(NSOrderedSet(array: state.pinnedSourceIDs.compactMap(canonicalLibrarySourceID))) as? [String] ?? [],
            selectedSourceID: canonicalLibrarySourceID(state.selectedSourceID),
            selectedSortBySource: dictionaryPreservingLastValue(
                state.selectedSortBySource.compactMap { key, value in
                    canonicalLibrarySourceID(key).map { ($0, value) }
                }
            ),
            selectedBankBySource: dictionaryPreservingLastValue(
                state.selectedBankBySource.compactMap { key, value in
                    canonicalLibrarySourceID(key).map { ($0, value) }
                }
            )
        )
    }
}

extension Notification.Name {
    static let pinballLibrarySourcesDidChange = Notification.Name("pinballLibrarySourcesDidChange")
}

func postPinballLibrarySourcesDidChange() {
    NotificationCenter.default.post(name: .pinballLibrarySourcesDidChange, object: nil)
}

func normalizeStringMap(_ raw: [String: Any]?) -> [String: String] {
    dictionaryPreservingLastValue((raw ?? [:]).compactMap { key, value in
        guard let canonicalKey = canonicalLibrarySourceID(key), let stringValue = value as? String else { return nil }
        return (canonicalKey, stringValue)
    })
}

func normalizeIntMap(_ raw: [String: Any]?) -> [String: Int] {
    dictionaryPreservingLastValue((raw ?? [:]).compactMap { key, value in
        guard let canonicalKey = canonicalLibrarySourceID(key) else { return nil }
        if let intValue = value as? Int { return (canonicalKey, intValue) }
        if let numberValue = value as? NSNumber { return (canonicalKey, numberValue.intValue) }
        return nil
    })
}

func dedupedPairs<Key: Hashable, Value>(_ pairs: [(Key, Value)]) -> [(Key, Value)] {
    Array(dictionaryPreservingLastValue(pairs))
}

func dictionaryPreservingLastValue<Key: Hashable, Value>(_ pairs: [(Key, Value)]) -> [Key: Value] {
    var out: [Key: Value] = [:]
    for (key, value) in pairs {
        out[key] = value
    }
    return out
}
